base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -22,97 +22,323 @@
     match monero_difficulty_calculation(header) {
-        Ok(v) => v,
-        Err(_) => 1.into(), // todo this needs to change to 0 when merge mine is implemented
+        Ok(d) => (d.0, d.1),
+        Err(_) => (1.into(), Hash::null_hash().0.to_vec()),
     }
 }
 
+pub fn merged_mining_subfield(header: &BlockHeader) -> SubField {
+    let tag = merged_mining_tag_hash(header);
+    let depth = 0;
+    SubField::MergeMining(VarInt(depth), tag)
+}
+
+fn merged_mining_transaction_hash(header: &BlockHeader) -> Hash {
+    let mut merge_mining_prehash: Vec<u8> = header.version.to_le_bytes().to_vec();
+    merge_mining_prehash.append(&mut header.height.to_le_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.prev_hash.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.timestamp.as_u64().to_le_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.output_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.range_proof_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.kernel_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.total_kernel_offset.as_bytes().to_vec());
+    Hash::hash(merge_mining_prehash.as_slice())
+}
+
+fn merged_mining_tag_hash(header: &BlockHeader) -> Hash {
+    merged_mining_transaction_hash(header)
+}
+
 /// Internal function to calculate the difficulty attained for the given block Deserialized the Monero header from the
 /// provided header
-fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, MergeMineError> {
-    let monero = MoneroData::new(header)?;
-    verify_header(&header, &monero)?;
+fn monero_difficulty_calculation(header: &BlockHeader) -> Result<(Difficulty, Vec<u8>), MergeMineError> {
     #[cfg(feature = "monero_merge_mining")]
     {
+        println!("Deserializing monero header");
+        let monero = MoneroData::new(header)?;
+        println!("Verifying header");
+        verify_header(&header, &monero)?;
         let flags = RandomXFlag::get_recommended_flags();
-        let key = monero.key.clone();
+        // we need db access, this should be checked as part of the concensus code
+        // check_randomx_key(&monero)?;
+        println!("Creating input blob");
         let input = create_input_blob(&monero)?;
-        let cache = RandomXCache::new(flags, &key)?;
+        println!("Creating cache");
+        let cache = RandomXCache::new(flags, &monero.seed_key.as_bytes())?;
+        println!("Creating dataset");
         let dataset = RandomXDataset::new(flags, &cache, 0)?;
+        println!("Creating VM");
+        // Todo revisit check achieved difficulty. I believe this is an expensive call. Thats the reason monero does not
+        // create a new seed hash every block.
         let vm = RandomXVM::new(flags, Some(&cache), Some(&dataset))?;
-        let hash = vm.calculate_hash(&input)?;
+        println!("Calculating hash");
```

Comment:
We need to revisit check achieved difficulty. I believe this is an expensive call. Thats the reason monero does not use a new seed hash every block.  They only change every 2000 blocks. This allows them to reuse the randomX vm

Reply:
Agreed, this is still outstanding in #2061.